### PR TITLE
CMake: Use -isystem for third party headers

### LINKED ISSUE #3827: gcc 4.8 ignores Wunused-variable pragma
I am getting many warnings like
```
/ssd/libs-candi/boost_1_60_0/include/boost/system/error_code.hpp:221:36: warning: ‘boost::system::posix_category’ defined but not used [-Wunused-variable]
     static const error_category &  posix_category = generic_category();
```
from boost even though the include is wrapped by ``DEAL_II_DISABLE_EXTRA_DIAGNOSTICS`` which contains 
```
_Pragma("GCC diagnostic ignored \"-Wunused-variable\"")          
```
This seems to be a bug in gcc: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=64079

Any suggestion on working around it?

I just learned about ``-isystem`` instead of ``-I`` (see http://stackoverflow.com/a/15053852). Maybe we should use this for external includes like boost?

### PR DESCRIPTION
**Pull request by tamiko**
Use -isystem for third party includes

fixes https://github.com/dealii/dealii/issues/5466

[deleted old TODO list]

fixes #3827

### CODE REVIEW

**Comment by tamiko** [comment]
@tjhei That should do the trick.

**Comment by davydden** [comment]
@tamiko out of curiosity, what that `SYSTEM` do exactly? Hopefully it does NOT assume that all third-party libs are in system folders.

**Comment by tjhei** [comment]
Is this something we should delay to after 8.5?

**Comment by tjhei** [comment]
> remove the DEAL_II_(ENABLE|DISABLE)_EXTRA_DIAGNOSTICS macro

We are using this macro in ASPECT, so removing it would break this. Maybe turning it into a no-op instead?

**Comment by tamiko** [comment]
> Is this something we should delay to after 8.5?

Yes. I hope I made that clear with a bazillion of labels :-D

> We are using this macro in ASPECT, so removing it would break this. Maybe turning it into a no-op instead?

We can also keep the macro.

What I would like to avoid is a proliferation of it being abused in the library.
For example, we can simply cut the number of disabled warnings back in size and remove all internal usage to guard includes (and keep the three occurrences for deprecated static object initialization).

**Comment by tamiko** [comment]
The one thing that slightly itches me is the fact that we inevitably might include deal.II's headers via `-isystem` in <b>user projects</b>. We cannot avoid this with the logic currently implemented in cmake.

Maybe this is a good thing because only diagnostics originating in user code will be shown. Maybe it is bad because we lose important diagnostics (but frankly, this should not happen if -isystem works as advertised).

**Comment by tamiko** [comment]
Is this still of interest?

**Comment by davydden** [comment]
> The one thing that slightly itches me is the fact that we inevitably might include deal.II's headers via -isystem in user projects. We cannot avoid this with the logic currently implemented in cmake.

i agree that it's not really a bad thing. As long as dealii is has no warnings, it's ok for user projects to show only warnings from user code.

**Comment by davydden** [comment]
@tamiko can you please rebase and fix conflicts? Then we can test it with `gcc 7.2.0` which breaks `DEAL_II_(ENABLE|DISABLE)_EXTRA_DIAGNOSTICS` anyway.

**Comment by tamiko** [comment]
@davydden I could not yet reproduce with gcc-7.2.0 (at least the patched version we have in Gentoo).
Do you happen to know what patches Ubuntu has applied on top of vanilla gcc-7.2.0?

With respect to this change - I think we shouldn't rush the `-isystem` change just to work around a single warning produced by gcc-7.2.0. I am still not confident about how much diagnostics cmake might accidentally disable.

**Comment by tamiko** [comment]
@davydden Updated the branch.

File: cmake/macros/macro_deal_ii_setup_target.cmake [hunk 1]
```diff
@@ -103,9 +103,8 @@ MACRO(DEAL_II_SETUP_TARGET _target)
     SET(_build "RELEASE")
   ENDIF()
 
-  SET_PROPERTY(TARGET ${_target} APPEND PROPERTY
-    INCLUDE_DIRECTORIES "${DEAL_II_INCLUDE_DIRS}"
-    )
+  TARGET_INCLUDE_DIRECTORIES(${_target} SYSTEM PRIVATE ${DEAL_II_INCLUDE_DIRS})
```
Comment (tamiko) [inline cmake/macros/macro_deal_ii_setup_target.cmake]: We have to check when `TARGET_INCLUDE_DIRECTORIES` was introduced.

File: include/deal.II/base/config.h.in [hunk 2]
```diff
@@ -325,61 +324,6 @@
...
-_Pragma("GCC diagnostic ignored \"-Wtautological-constant-out-of-range-compare\"") \
-_Pragma("GCC diagnostic ignored \"-Wtype-limits\"")                     \
-_Pragma("GCC diagnostic ignored \"-Wundef\"")                           \
-_Pragma("GCC diagnostic ignored \"-Wunused-but-set-parameter\"")        \
-_Pragma("GCC diagnostic ignored \"-Wunused-but-set-variable\"")         \
-_Pragma("GCC diagnostic ignored \"-Wunused-function\"")                 \
-_Pragma("GCC diagnostic ignored \"-Wunused-parameter\"")                \
-_Pragma("GCC diagnostic ignored \"-Wunused-private-field\"")            \
-_Pragma("GCC diagnostic ignored \"-Wunused-variable\"")                 \
-_Pragma("GCC diagnostic warning \"-Wpragmas\"")                   /*!*/
```
Comment (tamiko) [inline include/deal.II/base/config.h.in]: Honestly speaking, this macro got out of control. Maybe admitting defeat and simply using `-isystem` is the right call.

**Comment by davydden** [comment]
> I could not yet reproduce with gcc-7.2.0 (at least the patched version we have in Gentoo).
> Do you happen to know what patches Ubuntu has applied on top of vanilla gcc-7.2.0?

No idea, but I think @BenBrands see the same warnings with `gcc-7.2.0` compiled by Spack on Ubuntu 16.04 or alike. So if you want to reproduce it, that might be a way to go.

File: cmake/macros/macro_deal_ii_setup_target.cmake (hunk 1 above)
Comment (davydden) [reply cmake/macros/macro_deal_ii_setup_target.cmake]: according to https://stackoverflow.com/questions/20643368/cmake-different-include-directories-for-different-targets/20643487#20643487 -- 2.8.12

**Comment by tamiko** [comment]
@davydden Works on my machine (with gcc-7.3.0) - I don't get any warnings. Neither in the build directory, nor in examples configured as external project (e.g. step-33 with sacado includes).

File: include/deal.II/base/config.h.in (hunk 2 above)
Comment (drwells) [reply include/deal.II/base/config.h.in]: Are there any downsides to this? I don't think that most users care if some third party library has headers with warnings: there really is not much we can do about it.

**Comment by davydden** [comment]
@tamiko i can confirm that this fixes https://github.com/dealii/dealii/issues/5466 . Please squash your commits.

**Comment by davydden** [comment]
./run-tests

**Comment by davydden** [comment]
@tamiko examples `step-xy` and python wrappers report warnings. So `isystem` is not applied there, it seems.

**Comment by tjhei** [comment]
As said a while ago, we need to keep the macro around, otherwise we break downstream projects that use it like ASPECT.

**Comment by davydden** [comment]
@tjhei 

> As said a while ago, we need to keep the macro around, otherwise we break downstream projects that use it like ASPECT.

IMO 9.0 release is a good time to break this. Downstream projects can check deal.ii's version and work-aroudn this rather easily. Otherwise this means that one needs to drag those macros forever.

**Comment by drwells** [comment]
I am in favor of keeping the macro around for compatibility. We should at the very least still define the symbol even if it has not value.

I think we should keep the macro as-is for 9.0 and then (for 9.1) just `#define` it (so that user codes still compile). At least aspect, pi-domus, and deal2lkit all use this macro and it doesn't cost us anything to keep it in the header.

**Comment by tjhei** [comment]
We also use this macro for other cases, for example to disable deprecation warnings for a particular function. So I would like to keep such a macro around (with its definition!). Otherwise I would reimplement it in ASPECT.

**Comment by tamiko** [comment]
I simply removed the macro to check whether `-isystem` behaves as intended.
We can keep it for a version. In the long run, I would prefer to remove it.

@davydden The `DEAL_II_SETUP_TARGET` macro should also use `-isystem` now. I can check in detail. Can you send me a `detailed.log` (via mail) so that I can have a look at what goes wrong? (When you say `step-xy` produces warnings - is this during build, or as an external project?)

**Comment by davydden** [comment]
@tamiko: that was during dealii build. Same with python wrappers

**Comment by tamiko** [comment]
> @tamiko: that was during dealii build. Same with python wrappers 

I will have a look at the corresponding macros. I'll let you know if I
need the `detail.log` (for figuring out the include directory layout).

**Comment by davydden** [comment]
@tamiko are warnings in deal.II examples fixed now?

**Comment by davydden** [comment]
@tamiko the current state of this PR still produces warnings in dealii examples and python wrappers.

**Comment by tamiko** [comment]
> @tamiko the current state of this PR still produces warnings in dealii examples and python wrappers.

I had no luck reproducing these warnings so far. Can you please send me
a detailed.log so that I can get an idea what directory structure you
have?

**Comment by davydden** [comment]
>  Can you please send me a detailed.log so that I can get an idea what directory structure you have?

here you go
[detailed.log](https://github.com/dealii/dealii/files/1695934/detailed.log)

**Comment by tamiko** [comment]
@davydden Thanks. on it.

**Comment by masterleinad** [comment]
It seems we are missing to modify `macro_deal_ii_insource_setup_target.cmake ` as well. Applying the following, I don't get the warnings for the examples anymore.
```
diff --git a/cmake/macros/macro_deal_ii_insource_setup_target.cmake b/cmake/macros/macro_deal_ii_insource_setup_target.cmake
index 4c7efcb..1bfd408 100644
--- a/cmake/macros/macro_deal_ii_insource_setup_target.cmake
+++ b/cmake/macros/macro_deal_ii_insource_setup_target.cmake
@@ -36,9 +36,10 @@ MACRO(DEAL_II_INSOURCE_SETUP_TARGET _target _build)
       "${CMAKE_BINARY_DIR}/include"
       "${CMAKE_SOURCE_DIR}/include"
       ${DEAL_II_BUNDLED_INCLUDE_DIRS}
-      ${DEAL_II_INCLUDE_DIRS}
     )
 
+  TARGET_INCLUDE_DIRECTORIES(${_target} SYSTEM PRIVATE ${DEAL_II_INCLUDE_DIRS})
+
   IF(CMAKE_VERSION VERSION_LESS 3.9 OR CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
     SET_PROPERTY(TARGET ${_target} APPEND_STRING PROPERTY
       COMPILE_FLAGS " ${DEAL_II_CXX_FLAGS} ${DEAL_II_CXX_FLAGS_${_build}}"
```

**Comment by tamiko** [comment]
@masterleinad Good point. I have pushed that change.

Further, I have retained the macro itself, but removed all internal usage.

That should address the last bits. Let's whether we can merge now :-)

**Comment by davydden** [comment]
@tamiko @masterleinad examples are good now (no warnings), `python` bidings still give warnings. 

But since no-one so far objects this PR, a few devs already support it and the macro is retained, let's merge this and fix the remaining issues later.

**Comment by drwells** [comment]
RIP
`ENABLE_INTERIOR_DIAGNOSTICS`
2015 - 2018

**Pull request merged**